src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/DeriveHearingCentreHandler.java

```diff
@@ -0,0 +1,62 @@
+package uk.gov.hmcts.reform.bailcaseapi.domain.handlers.presubmit;
+
+import static java.util.Objects.requireNonNull;
+import static uk.gov.hmcts.reform.bailcaseapi.domain.entities.BailCaseFieldDefinition.HEARING_CENTRE;
+import static uk.gov.hmcts.reform.bailcaseapi.domain.entities.BailCaseFieldDefinition.IRC_NAME;
+import static uk.gov.hmcts.reform.bailcaseapi.domain.entities.BailCaseFieldDefinition.PRISON_NAME;
+
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.BailCase;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.HearingCentre;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.Event;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.bailcaseapi.domain.handlers.PreSubmitCallbackHandler;
+import uk.gov.hmcts.reform.bailcaseapi.domain.service.HearingCentreFinder;
+
+@Component
+public class DeriveHearingCentreHandler implements PreSubmitCallbackHandler<BailCase> {
+
+    private final HearingCentreFinder hearingCentreFinder;
+
+    public DeriveHearingCentreHandler(HearingCentreFinder hearingCentreFinder) {
+        this.hearingCentreFinder = hearingCentreFinder;
+    }
+
+    public boolean canHandle(PreSubmitCallbackStage callbackStage, Callback<BailCase> callback) {
+        requireNonNull(callbackStage, "callbackStage must not be null");
+        requireNonNull(callback, "callback must not be null");
+
+        return callbackStage == PreSubmitCallbackStage.ABOUT_TO_SUBMIT
+               && callback.getEvent() == Event.START_APPLICATION;
+    }
+
+    public PreSubmitCallbackResponse<BailCase> handle(PreSubmitCallbackStage callbackStage,
+                                                      Callback<BailCase> callback) {
+        if (!canHandle(callbackStage, callback)) {
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        final BailCase bailCase = callback.getCaseDetails().getCaseData();
+
+        if (bailCase.read(HEARING_CENTRE).isEmpty()) {
+
+            setHearingCentreFromDetentionFacilityName(bailCase);
+        }
+
+        return new PreSubmitCallbackResponse<>(bailCase);
+    }
+
+    public void setHearingCentreFromDetentionFacilityName(BailCase bailCase) {
+        final String prisonName = bailCase.read(PRISON_NAME, String.class).orElse("");
+
+        final String ircName = bailCase.read(IRC_NAME, String.class).orElse("");
+
+        String detentionFacilityName = !prisonName.isEmpty() ? prisonName : ircName;
```

Comment:
Can you put a check here - if both prison name and irc name are empty, it should throw RequiredFieldMissingException.

Reply:
Added in a check and a test for the exception. Have a look and see if it's alright